compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/test/iree_gpu_ops.mlir

```diff
@@ -92,3 +92,134 @@ func.func @tensor_barrier_multiple_inputs(%input: tensor<?xf16>) -> (tensor<?xf1
+  %c1 = arith.constant 1 : index
+  %result = scf.forall (%i) in (%c1) shared_outs(%out = %dest) -> (tensor<256xf32>) {
+    scf.forall.in_parallel {
+      iree_gpu.coalesced_gather_dma %indices, %source into %out
+        : tensor<1024xindex>, tensor<1024xf32>, tensor<256xf32> -> tensor<256xf32>
+    }
+  }
+  return %result : tensor<256xf32>
+}
+
+// CHECK-LABEL: func @coalesced_gather_dma_1d
+//  CHECK-SAME:   %[[INDICES:[A-Za-z0-9]+]]: tensor<1024xindex>
+//  CHECK-SAME:   %[[SOURCE:[A-Za-z0-9]+]]: tensor<1024xf32>
+//  CHECK-SAME:   %[[DEST:[A-Za-z0-9]+]]: tensor<256xf32>
+//       CHECK:   scf.forall
+//       CHECK:     scf.forall.in_parallel
+//       CHECK:       iree_gpu.coalesced_gather_dma %[[INDICES]], %[[SOURCE]] into %{{.+}} : tensor<1024xindex>, tensor<1024xf32>, tensor<256xf32> -> tensor<256xf32>
+
+// -----
+
+// Test coalesced_gather_dma within scf.forall with tensor.parallel_insert_slice
+// Using simpler dimensions to satisfy the verifier constraint
+func.func @coalesced_gather_dma_in_forall(%indices: tensor<1024x64xindex>, %source: tensor<2048x64xf32>, %dest: tensor<256x16xf32>) -> tensor<256x16xf32> {
+  %c4 = arith.constant 4 : index
+  %result = scf.forall (%i) in (%c4) shared_outs(%out = %dest) -> (tensor<256x16xf32>) {
+    %c64 = arith.constant 64 : index
+    %c256 = arith.constant 256 : index
+    %c0 = arith.constant 0 : index
+    %c4_idx = arith.constant 4 : index
+    %offset_idx = arith.muli %i, %c256 : index
+    %offset_dest = arith.muli %i, %c4_idx : index
+
+    // Extract slices for this iteration
+    // indices: 64x16 = 1024 elements, dest: 64x16xf32 = 1024 * 4 bytes = 4096 bytes, ratio = 4096/1024 = 4
+    %indices_slice = tensor.extract_slice %indices[%offset_dest, 0] [64, 16] [1, 1]
+      : tensor<1024x64xindex> to tensor<64x16xindex>
+
+    // Insert the result back into the output tensor
+    scf.forall.in_parallel {
+      // Simply use coalesced_gather_dma as the parallel combining op
+      iree_gpu.coalesced_gather_dma %indices_slice, %source into %out
+        : tensor<64x16xindex>, tensor<2048x64xf32>, tensor<256x16xf32> -> tensor<256x16xf32>
+    }
+  }
+  return %result : tensor<256x16xf32>
+}
+
+// CHECK-LABEL: func @coalesced_gather_dma_in_forall
+//  CHECK-SAME:   %[[INDICES:[A-Za-z0-9]+]]: tensor<1024x64xindex>
+//  CHECK-SAME:   %[[SOURCE:[A-Za-z0-9]+]]: tensor<2048x64xf32>
+//  CHECK-SAME:   %[[DEST:[A-Za-z0-9]+]]: tensor<256x16xf32>
+//       CHECK:   %[[C4:.+]] = arith.constant 4 : index
+//       CHECK:   %[[RESULT:.+]] = scf.forall (%[[I:.+]]) in (%[[C4]]) shared_outs(%[[OUT:.+]] = %[[DEST]])
+//       CHECK:     %[[INDICES_SLICE:.+]] = tensor.extract_slice %[[INDICES]]
+//       CHECK:     scf.forall.in_parallel
+//       CHECK:       iree_gpu.coalesced_gather_dma %[[INDICES_SLICE]], %[[SOURCE]] into %[[OUT]] : tensor<64x16xindex>, tensor<2048x64xf32>, tensor<256x16xf32> -> tensor<256x16xf32>
+//       CHECK:   return %[[RESULT]]
```

Comment:
Can we get a test that looks like the expected usage - that is, a workgroup forall containing a subgroup forall containing a DMA gather?

I'd like to make sure the IR we're trying to lower to gather_to_lds is something we all agree on.